contracts/validator/contract/XDCValidatorV2.sol

```diff
@@ -0,0 +1,518 @@
+pragma solidity ^0.4.26;
+
+import "./libs/SafeMath.sol";
+
+contract XDCValidator {
+    using SafeMath for uint256;
+
+    event Vote(address _voter, address _candidate, uint256 _cap);
+    event Unvote(address _voter, address _candidate, uint256 _cap);
+    event Propose(address _owner, address _candidate, uint256 _cap);
+    event Resign(address _owner, address _candidate);
+    event Withdraw(address _owner, uint256 _blockNumber, uint256 _cap);
+    event UploadedKYC(address _owner, string kycHash);
+    event InvalidatedNode(address _masternodeOwner, address[] _masternodes);
+
+    struct ValidatorState {
+        address owner;
+        bool isCandidate;
+        uint256 cap;
+        mapping(address => uint256) voters;
+    }
+
+    struct WithdrawState {
+        mapping(uint256 => uint256) caps;
+        uint256[] blockNumbers;
+    }
+
+    mapping(address => WithdrawState) withdrawsState;
+
+    mapping(address => ValidatorState) validatorsState;
+    mapping(address => address[]) voters;
+
+    // Mapping structures added for KYC feature.
+    mapping(address => string[]) public KYCString;
+    mapping(address => uint) public invalidKYCCount;
+    mapping(address => mapping(address => bool)) public hasVotedInvalid;
+    mapping(address => address[]) public ownerToCandidate;
+    address[] public owners;
+
+    address[] public candidates;
+
+    uint256 public candidateCount = 0;
+    uint256 public ownerCount = 0;
+    uint256 public minCandidateCap;
+    uint256 public minVoterCap;
+    uint256 public maxValidatorNumber;
+    uint256 public candidateWithdrawDelay;
+    uint256 public voterWithdrawDelay;
+
+    mapping(address => bool) public invalidOwner;
+
+    mapping(address => bool) public invalidCandidate;
```

Comment:
Does `invalidOwner` and `invalidCandidate` can be combined to `invalidOwnerOrCandidate` ?

Reply:
no, the different logic for different business , sometimes address is invalidCandidate, but it is normal owner, it can run some function , if set invalidOwnerOrCandidate, logic will be ban